Fix issue #20405: Add Duplicate answer group in the validation warning list, Display warning message for all interaction error

EliasBerhe opened the pull request:
## Overview

1. This PR fixes or fixes part of https://github.com/oppia/oppia/issues/20405.
2. This PR does the following:
    - Catches Duplicate answer group in the item selection interaction and adds it into validation warnings.
    - Validation warnings now indicate to the exploration author what is wrong and how to fix it for all interactions.
3. (For bug-fixing PRs only) The original bug occurred because: 
    - The original algorithm to catch duplicate answer group in the item selection interaction either wasn't working or wasn't adding it into to the list of validation warnings. Created a new algorithm to catch if there any duplicate group in that interaction and display it in the list of validation warnings.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

### Before

https://github.com/user-attachments/assets/6bdbde40-bb45-4064-b7f9-0059de739022

### After


https://github.com/user-attachments/assets/75bc33e2-134a-4f23-9152-2d593c9217ec


<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

EliasBerhe committed:
updated explorationWarning service to add all warning message

EliasBerhe committed:
modified some failed frontend test

EliasBerhe committed:
fixed duplicate answer group in itemSelectionInput

EliasBerhe committed:
Merge branch 'develop' into fixedDuplicateAnswerGroup

seanlip commented:
Assigning @amyyeung17 for shadow review. @amyyeung17 please feel free to reassign me when you believe it's ready to merge, thanks!

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,12 +319,42 @@ export class ItemSelectionInputValidationService {
         }
       });
     });
+    const ruleTypes = new Set([
+      'IsProperSubsetOf',
+      'Equals',
+      'ContainsAtLeastOneOf',
+    ]);
```
It seems like item-selection-input-rules.services.ts contains another rule. Is 'DoesNotContainAtLeastOneOf' ever considered? Sets are more for checking for membership since they enforce uniqueness and have quick access & delete time complexity.

EliasBerhe replied:
Yes! item-selection-input does contain another rule type- omit at least on of. I could just use object instead of listing them as sets because there is only fixed number of them. So, there is only 4 rule types for item-selection-input that I checked are used so I don't know if 'DoesNotContainAtLeastOneOf' is considered.

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,12 +319,42 @@ export class ItemSelectionInputValidationService {
+    ]);
+    const typeToInputMap = new Map<string, Set<string>>();
+    const ruleToAnswerGroup = new Map<string, number>();
+    ruleTypes.forEach(type => {
+      typeToInputMap.set(type, new Set());
+    });
```
Since we have a fixed number of rules, an object might be a better choice. We can directly instantiate it instead.

EliasBerhe replied:
I will try to use an object and reach back to you!

EliasBerhe replied:
If we did use object thought, inside the loop where we retrieve rule.type, do I then need to create an object for it? Can you explain how creating an object for these rule type would work? Sorry I was just a bit confused while trying to implement it.

amyyeung17 replied:
@EliasBerhe Please do not resolve comments ([Rules for making PRs](https://github.com/oppia/oppia/wiki/Rules-for-making-PRs#:~:text=Please%20do%20not%20mark%20the%20comment%20as%20resolved%2C%20since%20this%20just%20makes%20it%20harder%20to%20actually%20read%20the%20comment%20thread.)). 

We can simply create an object by instantiating it as const allRules = {}. There are several ways to set key and values, but the most basic way is allRules['rule-1'] = some value. We can traverse them similarly to an array with different syntax. A good starting point is [here.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,12 +319,42 @@ export class ItemSelectionInputValidationService {
+      for (let [ruleIndex, rule] of group.rules.entries()) {
+        const itemSelectionInputs =
+          rule.inputs as unknown as ItemSelectionRuleInputs;
+        const input = JSON.stringify([...itemSelectionInputs.x].sort());
```
Could you verify if the values in itemSelectionInputs.x are added in a certain order? And are all the values primitive? If so, we can simplify this statement to Array.toString().

EliasBerhe replied:
Values in the itemSelectionInput.x are added in order of in which they are created. So sorting it helps find duplicate answer groups. I first used Array.toString() however, it returned some weird value and couldn't stringify the array so I ended up using JSON.stringify which works as exepected.n

amyyeung17 replied:
Hm, I see. That makes sense. Side note, this works fine here, but sorting strings without callback can lead to funky behavior depending on what you want (ex. working with numbers and strings).

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,12 +319,42 @@ export class ItemSelectionInputValidationService {
+        const itemSelectionInputs =
+          rule.inputs as unknown as ItemSelectionRuleInputs;
+        const input = JSON.stringify([...itemSelectionInputs.x].sort());
+        const inputsSet = typeToInputMap.get(rule.type) || new Set();
```
If we are guaranteed to only encounter 3 (or 4 if verified) types, we can assume that we'll always get some value back.

EliasBerhe replied:
I don't exactly understand your question here. Do you mean when we call the get method on typeToInputMap (typeToInputMap.get(rule.type)? If so, if we are guaranteed that we encounter 4types, then because our keys in our map are those rule types it will give us a value. If not, we end up with empty set and it will keep the logic of the algorithm intact.

amyyeung17 replied:
Ah, it seems like I might have been slightly confused with the original logic.  DoesNotContainAtLeastOneOf is not supposed to add a warning. Got it.

Currently, you're checking to see the rule type is valid. If not, inputSet becomes to a new Set. We then check to see if the input exists in the potentially empty new Set. Is there a way that we can combine all of this logic so we don't have to create a new Set each time to represent that the rule is not valid? That way, we can avoid space overhead and an extra input check with invalid rule. 

In other words, how do we check for membership and retrieve its associated value?

EliasBerhe replied:
Yes we can just use if (!inputsSet) continue; thus instead of creating a new set we just skip past it.

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,12 +319,42 @@ export class ItemSelectionInputValidationService {
+          });
+        }
+
+        (typeToInputMap.get(rule.type) || new Set()).add(input);
```
Ditto here for line 343.

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,12 +319,42 @@ export class ItemSelectionInputValidationService {
+          warningsList.push({
+            type: AppConstants.WARNING_TYPES.ERROR,
+            message:
+              `The rule ${ruleIndex + 1} of answer group ` +
+              `${answerGroupIndex + 1} is already present in answer group ${(ruleToAnswerGroup.get(input) ?? 0) + 1} -- ` +
+              'please remove or edit the rule in the answer group to avoid duplicate rules',
+          });
```
Assuming this is for the same rule, what if there are multiple answer groups with the same answers? 

Are there any scenarios where there are several identical groups? For example, 1 & 2 both have A & B, and 3 & 4 have C & D.

EliasBerhe replied:
So, if multiple answer group have the same rule (which is the rule type that can be equalsTo and etc, and rule input which is the answer), then that is a duplicate and should be displayed. However, say 1 has rule such that its type is "equals to" and input (answer) is "a" and 2 has rule such that rule type is "contains at least one of" and input (answer) is "a" again then that is not a duplicate.

amyyeung17 replied:
I don't fully recall what the reasoning behind my comment was, whoops. Looking at it now though, does IsProperSubsetOf ever clash with Equals?

EliasBerhe replied:
No, isProperSubsetOf won't clash with Equals because they are two different rule types. Since algorithm catches duplicate answer per rule type,  Equals and isProperSubsetOF won't clash

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,12 +319,42 @@ export class ItemSelectionInputValidationService {
+      'Equals',
+      'ContainsAtLeastOneOf',
+    ]);
+    const typeToInputMap = new Map<string, Set<string>>();
```
Would it be possible to use a more descriptive name for these variables? I think it would help make the code a bit clearer.

EliasBerhe replied:
Ofcourse!

amyyeung17 replied:
Another point to mention. Maps are better for large, complex data (keys that are not strings) that is updated rather frequently. It considers the order of insertion too.

amyyeung17 reviewed (commented):
Hi @EliasBerhe, thanks for your work on this! Just to get the ball rolling, here are a few initial comments. This review will be done incrementally, so you'll receive more detailed feedback as we go.

EliasBerhe commented:
@amyyeung17  sorry for late response, I was out for a week on vacation. I have addressed most of your comments and I will edit some of the changes I need to make by tomorrow.

amyyeung17 commented:
Hi @EliasBerhe, please reassign me when you're ready for another review.

EliasBerhe commented:
@amyyeung17 PTAL
For another review! I am facing issues pushing. Until fixed, please make another review. Thanks!

EliasBerhe committed:
fixed package.json for pushing

EliasBerhe committed:
fixed files to adjust to merge conflict

EliasBerhe committed:
Merge branch 'fixedDuplicateAnswerGroup' of https://github.com/CAFE-Learning/oppia into fixedDuplicateAnswerGroup

EliasBerhe committed:
Merge branch 'develop' into fixedDuplicateAnswerGroup

EliasBerhe commented:
@seanlip @amyyeung17 PTAL 
This PR has been opened for quite sometime now. Thank you!

EliasBerhe committed:
switched to using objects instead of maps to store rules

EliasBerhe committed:
added curly to if condition to pass lint checks

EliasBerhe committed:
added typeCast to past typescript check

EliasBerhe committed:
defined object signatures for objects used

EliasBerhe committed:
fixed more linit issues

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -325,37 +325,37 @@ export class ItemSelectionInputValidationService {
         const input = JSON.stringify([...itemSelectionInputs.x].sort());
-        const inputsSet = typeToInputMap.get(rule.type) || new Set();
+        const ruleKey = `${rule.type}:${input}`;
+        const inputsSet = ruleInputPerRuleTypeMap[rule.type];
+        if (!inputsSet) {
+          continue;
+        }
```
Continue and breaks are for loops. This does not skip the following if statement (inputSett.has(input)). 

I suggest checking for the key using Object.hasOwn. You can read more about the nuances between key in obj, obj[key], and hasOwn (replaces hasOwnProperty) below:

- [HasOwn](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn)
- [Downsides of checking obj[key]](https://stackoverflow.com/questions/66309979/js-checking-object-with-key-vs-hasownproperty)
- [key in obj vs. hasOwn (hasOwnProperty)](https://stackoverflow.com/questions/13632999/ifkey-in-object-or-ifobject-hasownpropertykey)

EliasBerhe replied:
I have changed it to use hasOwnProperty instead!

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -325,37 +325,37 @@ export class ItemSelectionInputValidationService {
+    const ruleInputPerRuleTypeMap: {[key: string]: Set<string>} = {};
+    const rulesPreAnswerGroup: {[key: string]: number} = {};
+    allRuleTypes.forEach((type: string) => {
+      ruleInputPerRuleTypeMap[type] = new Set();
```
Would it be possible to set the object values directly instead of declaring a separate set and using a for loop?

EliasBerhe replied:
Yes that would be possible, we would just need to add more lines

amyyeung17 replied:
Direct assignment works well in this case because we’re only using a few specific keys. It simplifies the code and improves readability, making it easier to understand and maintain.

amyyeung17 commented:
Hi @EliasBerhe, few more comments.

EliasBerhe committed:
directly instantiate objects and use hasOwnProperty instead on the objects

EliasBerhe commented:
@amyyeung17 all comments addressed!

amyyeung17 commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,39 @@ export class ItemSelectionInputValidationService {
+        const input = JSON.stringify([...itemSelectionInputs.x].sort());
+        const ruleKey = `${rule.type}:${input}`;
+        if (!ruleInputPerRuleTypeMap.hasOwnProperty(rule.type)) {
+          continue;
```
@EliasBerhe Instead of checking if it doesn't exist and continuing, could we run the logic only if it does? Additionally, we can chain the logic (since they’re relevant) into a single if statement using short-circuiting. Also, please reassign me when in the sidebar when you want me to take a look again. Thanks!

EliasBerhe replied:
Yap, that can be something we can do. I don't think I have the permission to assign you directly, however I can just assign you using PTAL command.

EliasBerhe replied:
Regarding chaining the logic into one if condition, I do believe it would be easier to understand if we use nested if instead. One if to check if ruleInputPerRuleTypeMap.hasOwnProperty(rule.type) and the second if inputSet.has(input) because the else condiition is relevant specifically if inputSet has the input or not. 
else {
            inputsSet.add(input);  -> HERE
            rulesPreAnswerGroup[ruleKey] = answerGroupIndex;
          }

amyyeung17 replied:
@EliasBerhe Ah sorry, my bad. I totally forgot that you can't assign. Thanks for bringing up that point. 

What about if (ruleInputPerRuleTypeMap.hasOwnProperty(rule.type) &&  ruleInputPerRuleTypeMap.get(rule.type).has(input))? Since checking for the key and then accessing the set are tightly coupled, combining them in one if statement keeps the logic direct. Nesting would only make sense if they weren’t so closely related. You have to check if the key exists before accessing it anyway.

EliasBerhe replied:
@amyyeung17 The was first thing I thought of when reading your comment. However, notice we have an else statement that adds the input to the inputSet. This would be done if the inputSet doesnot have the input (i.e ruleInputPerRuleTypeMap.get(rule.type).has(input) is false). 

So, if we combine, (ruleInputPerRuleTypeMap.hasOwnProperty(rule.type) && ruleInputPerRuleTypeMap.get(rule.type).has(input) into one if, we have to address somehow if the key exist (i.e ruleInputPerRuleTypeMap.hasOwnProperty(rule.type) is true ) but set doesn't have the input (i.e ruleInputPerRuleTypeMap.get(rule.type).has(input)  is false) so that the else statements logic can be valid.

amyyeung17 replied:
@EliasBerhe Ohhh, I see what you're saying. That's my bad. Would it be possible to avoid continue? 
```
if(ruleInputPerRuleTypeMap.hasOwnProperty(rule.type)) {
  const inputSet = ruleInputPerRuleTypeMap[rule.type];
  if(inputSet.has(input)) {
    ......
  } else {
    ......
  }
}
```

EliasBerhe replied:
@amyyeung17 Yes, the recent commit that I pushed reflects those changes.

EliasBerhe committed:
chain every condition to a single if

EliasBerhe committed:
changed to a two if conidtions

EliasBerhe commented:
@amyyeung17 PTAL. Can you take a look again? I think i have addressed all comments

EliasBerhe commented:
@amyyeung17 Let me know if there is anything that needs to be addressed!

amyyeung17 commented on core/templates/pages/exploration-editor-page/services/exploration-warnings.service.ts:
```diff
@@ -538,17 +538,18 @@ export class ExplorationWarningsService {
+        const formattedValue = Array.isArray(value)
+          ? value
+              .map(str => (str.endsWith('.') ? str.slice(0, -1) : str))
+              .join('; ') + '.'
```
@EliasBerhe I'm curious—was there a specific reason for deleting all the periods and manually re-adding them in this line? I think this is the last thing I wanted clarify before passing it to @seanlip

EliasBerhe replied:
@amyyeung17 Yea so we shouldn't be removing the last period using here
.map(str => (str.endsWith('.') ? str.slice(0, -1) : str)) , because we already manually deleted all those periods by going through all the files that contain those strings messages. But after we concatenate all those string using semicolon as delimeter, we then add period at the end. So I should remove the part where we removing the periods using str.slice(0, -1) but then after joining all those strings we should add the final period at the end.

amyyeung17 replied:
Sorry @EliasBerhe, my question about removing periods was about you deleting them from strings, especially the ones that aren't related to the answer groups (ex. line 83 where 'The first card of the lesson must be a checkpoint.' in app.constants.ts). 

I get that semicolons can join related sentences, but I think whether to use them over periods is a bit subjective. Also, the first word after a semicolon shouldn’t be capitalized unless it’s a proper noun as it currently is in your video.

EliasBerhe replied:
@amyyeung17 I understand your question now. So, previously we didn't delete all the periods and we just joined all the strings with a semicolon. So the idea was to go through all interaction warning and display both which interaction has errors and display the warning message too. However, @seanlip requested that we remove all the periods of string from a single interaction because it will confusing having periods and semicolons at the sametime. Example, if card A has some errors it will display like; "First Message.; Second Message.; Third Message.;". It would be less confusing if we just don't include the periods at all and just add them at the end.  You can see the conversation we had in our previous PR (which is closed due other issues ) https://github.com/oppia/oppia/pull/21920#discussion_r1956364042.  And also the issue ticket https://github.com/oppia/oppia/issues/20405#issuecomment-2638861991

amyyeung17 replied:
Oh I see. I thought it was slightly odd to replace the period with semicolons out of no where but this clarifies everything. Thanks for explaining @EliasBerhe.

EliasBerhe committed:
removed the logic that deletes the period at the end of each string in exploration service class

EliasBerhe commented:
@amyyeung17 last comment addressed!

amyyeung17 commented:
I think it LGTM, so passing it on to @seanlip. Thanks!

EliasBerhe commented:
@amyyeung17 Thank you for your extensive review!. @seanlip PTAL

seanlip commented:
@EliasBerhe Just a note, it'll take me a couple days to look at this (I'm quite slammed atm). Sorry for the delay -- will get to it soon!

seanlip commented on core/templates/pages/exploration-editor-page/services/exploration-warnings.service.ts:
```diff
@@ -538,17 +538,16 @@ export class ExplorationWarningsService {
-          Object.keys(this.stateWarnings).join(', ') +
-          '.',
-      });
+      for (const [key, value] of Object.entries(this.stateWarnings)) {
```
This seems a bit off to me. The values should all be arrays; single-warning cases should provide a single-element array. Then you do not need to do the "is it an array" checking here and can handle all of them uniformly.

EliasBerhe replied:
Yes! the values should all be arrays, it was just precautionary check but I would remove it so it can all done uniformly

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,38 @@ export class ItemSelectionInputValidationService {
     });
-
+    const ruleInputPerRuleTypeMap: {[key: string]: Set<string>} = {
+      IsProperSubsetOf: new Set<string>(),
```
Instead of hardcoding here, can you determine them programmatically (e.g. from the interactions json file that lists all the rule types)? Otherwise, if we add a new rule, devs may forget to update this.

Also TBH I don't understand what the name `ruleInputPerRuleTypeMap` means. Can you clarify? A comment might be helpful too.

EliasBerhe replied:
Yes thats a better idea. I have used the rule_template.json file to programmatically determince the ruleTypes. I will also rename my variables

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,38 @@ export class ItemSelectionInputValidationService {
+      Equals: new Set<string>(),
+      ContainsAtLeastOneOf: new Set<string>(),
+    };
+    const rulesPreAnswerGroup: {[key: string]: number} = {};
```
I don't know what `rulesPreAnswerGroup` means. Can you pick a better name, and possibly add a comment to explain? It should be possible for a developer to understand what this is storing without reading the implementation that follows.

EliasBerhe replied:
Yes I will rename all my variables and add comments to further explain their role

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,38 @@ export class ItemSelectionInputValidationService {
+          rule.inputs as unknown as ItemSelectionRuleInputs;
+        const input = JSON.stringify([...itemSelectionInputs.x].sort());
+        const ruleKey = `${rule.type}:${input}`;
+        if (ruleInputPerRuleTypeMap.hasOwnProperty(rule.type)) {
```
Why wouldn't this object have the rule, given that you've initialized it with each of the rule types?

EliasBerhe replied:
Yes it would have the rule.type its just precautionary check, I have removed it now though

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,38 @@ export class ItemSelectionInputValidationService {
+      for (let [ruleIndex, rule] of group.rules.entries()) {
+        const itemSelectionInputs =
+          rule.inputs as unknown as ItemSelectionRuleInputs;
+        const input = JSON.stringify([...itemSelectionInputs.x].sort());
```
This has a built-in assumption that all rules only have a single input `x`, which may be currently true but may not always be in the future. Can you find a more general way to serialize `inputs` reversibly?

EliasBerhe replied:
I have added logic that will serialize inputs reversibly if we later time use multiple inputs for an interaction

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,38 @@ export class ItemSelectionInputValidationService {
+        const input = JSON.stringify([...itemSelectionInputs.x].sort());
+        const ruleKey = `${rule.type}:${input}`;
+        if (ruleInputPerRuleTypeMap.hasOwnProperty(rule.type)) {
+          const inputsSet = ruleInputPerRuleTypeMap[rule.type];
```
seenInputs (or similar) conveys more semantically what this represents. You'll want to convey some idea of "this is what we've seen already", not just "this contains some inputs".

EliasBerhe replied:
Yes, that is a better variable name, I am trying rename all my variables for the next push

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,38 @@ export class ItemSelectionInputValidationService {
+        const ruleKey = `${rule.type}:${input}`;
+        if (ruleInputPerRuleTypeMap.hasOwnProperty(rule.type)) {
+          const inputsSet = ruleInputPerRuleTypeMap[rule.type];
+          if (inputsSet.has(input)) {
```
A better data structure might be a mapping from serialized input to the answer group index, instead of a set. Then you don't need the separate rulesPreAnswerGroup object.

EliasBerhe replied:
@seanlip I have question on that. If we map serialized inputs to the answer group index it would simplify the algorithm however, would that mean same inputs for different rule type would create redundancy? For example. for rule.type = Equals if our input say is A. If we create another rule where rule.type = ContainsAtleastOneOf and our input is A. is that duplicate? Because the way I was thinking was, we trying catch duplicates rules.inputs in the same rule.type i.e same inputs in different rule types are not redundant.

seanlip replied:
No, the case you are mentioning isn't a duplicate.

And ah, sorry, you are right, it might mean you need a mapping from the serialized "rule + input" to the answer group index, instead.

EliasBerhe replied:
Yes! I think that makes the whole thing much simpler. I have made the necessary change!

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,38 @@ export class ItemSelectionInputValidationService {
+          if (inputsSet.has(input)) {
+            warningsList.push({
+              type: AppConstants.WARNING_TYPES.ERROR,
+              message: `The rule ${ruleIndex + 1} of answer group ${answerGroupIndex + 1} is already present in answer group ${(rulesPreAnswerGroup[ruleKey] ?? 0) + 1} -- please remove or edit the rule in the answer group to avoid duplicate rules`,
```
Does this correctly handle the case where there are duplicate rules in the same answer group -- is there a test for that?

EliasBerhe replied:
Yes, it would handle duplicate rules for the same answer group, however I don't think I have test case for such scenario. But the idea is that, for each answer group it goes through all its rule and check redundancy. We can see the part where we are pushing warning is inside the double loop.

seanlip replied:
OK, please add a test case for this, just so we ensure the correct behaviour happens now and continues to happen in the future. Thanks.

EliasBerhe replied:
I have added test case for duplicate rule in the same and different answer group

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.spec.ts:
```diff
@@ -840,8 +840,7 @@ describe('ItemSelectionInputValidationService', () => {
         message:
-          'The rule 1 of answer group 0 of ItemSelectionInput interaction ' +
-          'is a duplicate.',
+          'The rule 2 of answer group 1 is already present in answer group 1 -- please remove or edit the rule in the answer group to avoid duplicate rules',
```
The rule 2 --> Rule 2

Good idea to change from 0-based indexing to 1-based indexing.

EliasBerhe replied:
done

seanlip commented on extensions/interactions/NumericInput/directives/numeric-input-validation.service.spec.ts:
```diff
@@ -296,8 +296,7 @@ describe('NumericInputValidationService', () => {
         message:
-          'Learner answer 1 input should be greater than ' +
-          'or equal to zero.',
+          'Learner answer 1 input should be greater than ' + 'or equal to zero',
```
If you are putting two strings in the same line, please drop the '+' between them. 

Please also fix this for the other cases in this PR.

EliasBerhe replied:
done

seanlip replied:
@EliasBerhe Why do you say this is done when it is not done?

EliasBerhe replied:
I was very confident I changed it! I don't know I must have not saved the file but I genuinely thought I resolved it. Regardless, I would fix and push again! Apologies for the mistake

EliasBerhe replied:
@seanlip I have went back and changed the file (Just dropped the string to then next line after +) Oddly, whenever I committed it, it reverts to way it shown here. I don't know why it is doing it but I have tried different way so to resolve it however it just goes back to the way it is. I have now just made it into one string (put the string inside one qoute) since the string itself is not too long.

EliasBerhe replied:
done

seanlip reviewed (changes requested):
Thanks @EliasBerhe, sorry for the delay in reviewing. Took a full pass, PTAL.

seanlip commented:
Also adding @lkbhitesh07 and @Vir-8 -- PTAL for codeowners. Thanks!

EliasBerhe committed:
removed array check for exploration warning list array

Vir-8 reviewed (approved):
LGTM for codeowners

EliasBerhe committed:
used different data structure to store rules to answer group and added test case for duplicate answer group in different group

EliasBerhe committed:
fixed linting issues

EliasBerhe committed:
fixed comments

EliasBerhe commented:
@seanlip PTAL

seanlip commented:
Hi @EliasBerhe: please could you reply to all comments, following the instructions in step 5 [here](https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve)? In particular, for comments that you've addressed, please reply "Done." -- this helps keep track of which comments have been addressed and which haven't. Thanks!

EliasBerhe committed:
drop strings after + in test files

EliasBerhe committed:
fix additional linting problems

EliasBerhe committed:
change variable names

EliasBerhe committed:
remove trailing space in number input validation test class

EliasBerhe committed:
Used capital R Rule instead of rule for duplicate answer group message

EliasBerhe commented:
@seanlip Sorry for that! I believe I have now addressed all comments and left "done" . PTAL

EliasBerhe committed:
fix rever commit issues

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,39 @@ export class ItemSelectionInputValidationService {
     });
-
+    // Map serialized rule + input to answer group index.
+    const mapRulesToAnswerGroupIndex: {[key: string]: number} = {};
```
We avoid starting variable names with verbs. Perhaps `ruleToAnswerGroupIndex` or `ruleToAnswerGroupMapping`?

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,39 @@ export class ItemSelectionInputValidationService {
+          const ag = mapRulesToAnswerGroupIndex[ruleKey];
+          warningsList.push({
+            type: AppConstants.WARNING_TYPES.ERROR,
+            message: `The Rule ${ruleIndex + 1} of answer group ${answerGroupIndex + 1} is already present in answer group ${ag} -- please remove or edit the rule in the answer group to avoid duplicate rules`,
```
"The Rule" --> "Rule"

Note that I already commented on this previously -- see https://github.com/oppia/oppia/pull/22062#discussion_r2042986823 . Please try to address comments fully.

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,39 @@ export class ItemSelectionInputValidationService {
+            [...value].sort(),
+          ])
+        );
+        const input = JSON.stringify(allInputs);
```
This variable name is misleading -- you aren't converting all the inputs to just one input. Use something like `stringifiedInputs`, instead.

EliasBerhe replied:
done

seanlip commented on extensions/interactions/ItemSelectionInput/directives/item-selection-input-validation.service.ts:
```diff
@@ -343,13 +319,39 @@ export class ItemSelectionInputValidationService {
+        const allInputs = Object.fromEntries(
+          Object.entries(itemSelectionInputs).map(([key, value]) => [
+            key,
+            [...value].sort(),
```
I am not sure this is general enough, you are assuming that each input value is a list. What if there was a rule that's something like "matches answer x but with at most N wrong", where N is an integer-type input?

EliasBerhe replied:
I've modified the code to first check if each value is an array (using Array.isArray()) before applying sorting. For other types, I have kept it as it is.

EliasBerhe replied:
done

seanlip reviewed (changes requested):
Thanks @EliasBerhe, it definitely looks better than before. All my remaining comments are in the item-selection-input-valication TS file, PTAL.

EliasBerhe committed:
fix lint issues failing CI/CD

EliasBerhe committed:
changes inputs to stringfied inputs

EliasBerhe committed:
fix tests failing in item-selection validation spec class

EliasBerhe committed:
Genearlized serialization of rule inputs to handle different data types

EliasBerhe committed:
addressed linting issue

EliasBerhe committed:
catch unused variable in item-selection validation class as part of liniting

EliasBerhe commented:
@seanlip PTAL

seanlip reviewed (changes requested):
Just one comment.

EliasBerhe commented:
@seanlip  I was very confident I changed it! I don't know I must have not saved the file but I genuinely thought I resolved it. Regardless, I would fix and push again! Apologies for the mistake

EliasBerhe committed:
fix concatenation of strings by dropping strings after +

EliasBerhe committed:
add a space after + on some test files

EliasBerhe committed:
remove full stops from number-input-validation spec class from front-end test pass

EliasBerhe committed:
fix test case for greater than and equal to zero liniting issues

EliasBerhe committed:
Fix again liniting issues in numeric input test file

EliasBerhe committed:
fix liniting issues

EliasBerhe committed:
check commit if reverts back

EliasBerhe committed:
check again if it reverts back after commit

EliasBerhe committed:
final fix to numer-input validation test class after changes revert after commiting

EliasBerhe committed:
fix front-end test failing in the number-input validation class

EliasBerhe committed:
removed + in test case in numeric input validation to see if commit works

EliasBerhe commented:
@seanlip PTAL

seanlip reviewed (approved):
Thanks @EliasBerhe, LGTM!

Also, just for reference, I think the auto-formatter pushes it to the same line precisely because there's no need for the break (since the string is short enough). So, concatenating is the way to go.

EliasBerhe committed:
Merge branch 'develop' into fixedDuplicateAnswerGroup

EliasBerhe commented:
@lkbhitesh07 PTAL

lkbhitesh07 reviewed (approved):
LGTM! Thanks

EliasBerhe committed:
Merge branch 'develop' into fixedDuplicateAnswerGroup

EliasBerhe committed:
fix failing front-end tests in item-selection-input validation test class

EliasBerhe committed:
add more front-end coverage test to itemSelection input class

EliasBerhe committed:
fix test class in inputSelectionValidation class by sorting keys in a object

EliasBerhe committed:
Merge branch 'develop' into fixedDuplicateAnswerGroup

EliasBerhe committed:
pull final changes:

EliasBerhe commented:
@seanlip All CI check pass. I was hoping this can be merged. Thank you!

EliasBerhe committed:
Merge branch 'develop' into fixedDuplicateAnswerGroup

seanlip commented:
@EliasBerhe -- so sorry, we are having a bit of a problem with our CI in develop, and @jayam04 is fixing this in https://github.com/oppia/oppia/pull/22663. I think you can only merge this PR once that goes in. Apologies, hopefully this will get sorted within the next 24 hours.
